### PR TITLE
JAMES-3117 More Reactive healthChecks

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
`RabbitMQEventBusTest$LifeCycleTest$SingleEventBus.dispatchShouldWorkAfterRestartForNewKeyRegistration » IllegalState`

test this please

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java [hunk 1]
```diff
@@ -46,27 +49,44 @@ public ComponentName componentName() {
...
 
     @Override
-    public Result check() {
+    public Mono<Result> checkReactive() {
         try {
-            if (connectionPool.tryConnection() && rabbitChannelPoolImpl.tryChannel()) {
-                Optional<RabbitMQServerVersion> version = connectionPool.version();
+            return Flux.concat(connectionPool.tryConnection(),
+                rabbitChannelPoolImpl.tryChannel())
+                .reduce(true, (a, b) -> a && b)
```
Comment (mbaechler) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java:56]: what about `.filter(status -> status == true).then(checkVersion()).switchIfEmpty(unhealthy)`?

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/ReactorRabbitMQChannelPool.java [hunk 2]
```diff
@@ -214,18 +214,16 @@ public void close() {
...
-        } finally {
-            if (channel != null) {
-                borrowedChannels.remove(channel);
-                pool.returnObject(channel);
-            }
-        }
+    public Mono<Boolean> tryChannel() {
+        return Mono.usingWhen(borrow(),
+            channel -> Mono.just(channel.isOpen()),
+            channel -> {
```
Comment (mbaechler) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/ReactorRabbitMQChannelPool.java:220]: why a null check here and not on `isOpen()`?

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java [hunk 3]
```diff
@@ -46,27 +49,44 @@ public ComponentName componentName() {
...
+                    "Unhealthy RabbitMQ instances: could not establish a connection", e)));
+        } catch (Exception e) {
+            return Mono.just(Result.unhealthy(COMPONENT_NAME,
+                "Unhealthy RabbitMQ instances: could not establish a connection", e));
+        }
+    }
+
+    private Mono<? extends Result> checkVersion() {
+        return connectionPool.version()
+            .map(Optional::of)
```
Comment (mbaechler) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java:74]: why do we need some Optional here, can't we handle things with Mono?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchHealthCheck.java [hunk 4]
```diff
@@ -54,20 +55,16 @@ public ComponentName componentName() {
...
-            ClusterHealthResponse response = client.cluster()
-                .health(request, RequestOptions.DEFAULT);
-
-            return toHealthCheckResult(response);
-        } catch (IOException e) {
-            return Result.unhealthy(COMPONENT_NAME, "Error while contacting cluster", e);
-        }
+        return client.health(request)
+            .map(this::toHealthCheckResult)
+            .onErrorResume(IOException.class, e -> Mono.just(
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchHealthCheck.java:66]: what about runtime exception?

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java (hunk 1 above)
Comment (chibenwa) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java:56]: I prefer the reduce version.

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/ReactorRabbitMQChannelPool.java (hunk 2 above)
Comment (chibenwa) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/ReactorRabbitMQChannelPool.java:220]: It's a mono wrapping as the managed resource is an Publisher, it's required for typing.

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java (hunk 3 above)
Comment (chibenwa) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java:74]: No as we need the version and if it is acceptable for the error message.

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchHealthCheck.java (hunk 4 above)
Comment (chibenwa) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchHealthCheck.java:66]: Handled by callers.

**Comment by chibenwa** [comment]
Can be considered :green_apple:

File: backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java (hunk 1 above)
Comment (mbaechler) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backends/rabbitmq/RabbitMQHealthCheck.java:56]: I know that you do, still I think it's overkill.
But do as you want.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**